ci/remote-watch.py

```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/env python
+
+"""
+This command must be run in a git repository.
+
+It watches the remote branch for changes, killing the given PID when it detects
+that the remote branch no longer points to the local commit.
+
+If the commit message contains a line saying "CI_KEEP_ALIVE", then killing
+will not occur until the branch is deleted from the remote.
+
+If no PID is given, then the entire process group of this process is killed.
+"""
+
+# Prefer to keep this file Python 2-compatible so that it can easily run early
+# in the CI process on any system.
+
+import argparse
+import errno
+import logging
+import os
+import signal
+import subprocess
+import sys
+import time
+
+logger = logging.getLogger(__name__)
+
+GITHUB = "GitHub"
+TRAVIS = "Travis"
+
+
+def git(*args):
+    cmdline = ["git"] + list(args)
+    return subprocess.check_output(cmdline).decode("utf-8").rstrip()
+
+
+def get_current_ci():
+    if "GITHUB_WORKFLOW" in os.environ:
+        return GITHUB
+    elif "TRAVIS" in os.environ:
+        return TRAVIS
+    return None
+
+
+def get_ci_event_name():
+    ci = get_current_ci()
+    if ci == GITHUB:
+        return os.environ["GITHUB_EVENT_NAME"]
+    elif ci == TRAVIS:
+        return os.environ["TRAVIS_EVENT_TYPE"]
+    return None
+
+
+def get_repo_slug():
+    ci = get_current_ci()
+    if ci == GITHUB:
+        return os.environ["GITHUB_REPOSITORY"]
+    elif ci == TRAVIS:
+        return os.environ["TRAVIS_REPO_SLUG"]
+    return None
+
+
+def get_remote_url(remote):
+    return git("ls-remote", "--get-url", remote)
+
+
+def git_remote_branch_info():
+    """Obtains the remote branch name, remote name, and commit hash that
+    correspond to the local HEAD.
+
+    Returns:
+        ("refs/heads/mybranch", "origin", "1A2B3C4...")
+    """
+    ref = None
+    remote = None
+    expected_sha = None
+
+    try:
+        # Try to get the local branch ref. (e.g. refs/heads/mybranch)
+        head = git("symbolic-ref", "-q", "HEAD")
+        # Try to get the remotely tracked ref, if any. (e.g. origin/mybranch)
+        ref = git("for-each-ref", "--format=%(upstream:short)", head)
+    except subprocess.CalledProcessError:
+        pass
+
+    if ref:
+        (remote, ref) = ref.split("/", 1)
+        ref = "refs/heads/" + ref
+    else:
+        remote = git("remote", "show", "-n").splitlines()[0]
+        ref = os.getenv("TRAVIS_PULL_REQUEST_BRANCH")
+
+        if ref:
+            TRAVIS_PULL_REQUEST = os.environ["TRAVIS_PULL_REQUEST"]
+            if os.getenv("TRAVIS_EVENT_TYPE") == "pull_request":
+                ref = "refs/pull/{}/merge".format(TRAVIS_PULL_REQUEST)
+            else:
+                ref = "refs/heads/{}".format(TRAVIS_PULL_REQUEST)
+            expected_sha = os.getenv("TRAVIS_COMMIT")
+        else:
+            ref = os.getenv("GITHUB_REF")
```

Comment:
`TRAVIS_PULL_REQUEST_BRANCH` is not used as a ref here. It seems we just need to check whether we are in travis or not? I think the following makes a bit more clear, please correct me.

```suggestion
        if os.getenv("TRAVIS"):
            pr_number = os.environ["TRAVIS_PULL_REQUEST"]
            if os.getenv("TRAVIS_EVENT_TYPE") == "pull_request":
                ref = "refs/pull/{}/merge".format(pr_number)
            else:
                ref = "refs/heads/{}".format(pr_number)
            expected_sha = os.getenv("TRAVIS_COMMIT")
        else:
            ref = os.getenv("GITHUB_REF")
```

Reply:
Being in Travis doesn't imply it's a PR, so these aren't equivalent in theory. I'm not sure if it'll behave correctly in practice.